### PR TITLE
Specify recoverWithNull when reading JSON files

### PR DESCRIPTION
**Pull request by andygrove**
Closes https://github.com/NVIDIA/spark-rapids/issues/9326
Closes https://github.com/NVIDIA/spark-rapids/issues/9310

Depends on https://github.com/rapidsai/cudf/pull/14279

This PR enables the new cuDF option `recoverWithNulls` (see https://github.com/rapidsai/cudf/issues/12532) when reading from JSON files.

### CODE REVIEW

**Comment by andygrove** [comment]
build

File: sql-plugin/src/main/scala/org/apache/spark/sql/catalyst/json/rapids/GpuJsonScan.scala [hunk 1]
```diff
@@ -267,7 +267,14 @@ object JsonPartitionReader {
       RmmRapidsRetryIterator.withRetryNoSplit(dataBufferer.getBufferAndRelease) { dataBuffer =>
         withResource(new NvtxWithMetrics(formatName + " decode",
           NvtxColor.DARK_GREEN, decodeTime)) { _ =>
-          Table.readJSON(cudfSchema, jsonOpts, dataBuffer, 0, dataSize)
+          try {
+            Table.readJSON(cudfSchema, jsonOpts, dataBuffer, 0, dataSize)
+          } catch {
+            case e: AssertionError if e.getMessage == "CudfColumns can't be null or empty" =>
```
Comment (jlowe) [inline sql-plugin/src/main/scala/org/apache/spark/sql/catalyst/json/rapids/GpuJsonScan.scala:273]: Parsing exception messages is an anti-pattern.  Would be nice to have a followup to get a more specific exception type to help distinguish this case.

**Review by jlowe — APPROVED** [review]

File: sql-plugin/src/main/scala/org/apache/spark/sql/catalyst/json/rapids/GpuJsonScan.scala (hunk 1 above)
Comment (revans2) [reply sql-plugin/src/main/scala/org/apache/spark/sql/catalyst/json/rapids/GpuJsonScan.scala:273]: I agree, probably would be better to do this for all Exceptions and AssertionErrors.

Comment (andygrove) [reply sql-plugin/src/main/scala/org/apache/spark/sql/catalyst/json/rapids/GpuJsonScan.scala:273]: Thanks. I filed https://github.com/rapidsai/cudf/issues/14331

**Pull request merged**